pkg/ruler/api.go

```diff
@@ -500,9 +500,31 @@ func (a *API) ListRules(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 	if len(missing) > 0 {
-		// This API is expected to be strongly consistent, so it's an error if any rule group was missing.
-		http.Error(w, fmt.Sprintf("an error occurred while loading %d rule groups", len(missing)), http.StatusInternalServerError)
-		return
+		// This API is expected to be strongly consistent, but we expect the object storage to be strongly
+		// consistent too. This means that if a rule group existed when we listed the storage but doesn't exist
+		// (it's missing) when we load it, then it could have been deleted in the meanwhile.
+		//
+		// We don't want to consider this condition a failure, so we don't return an error in this case, but we
+		// just log a warning because it could be of interest for an operator to further investigate it.
+		level.Warn(a.logger).Log(
+			"msg", "list rules API skipped some rule groups, because missing when loading them after listing the storage (this could be due to rule groups deleted between listing the storage and getting rule groups content)",
+			"listed_rule_groups", len(rgs),
+			"missing_rule_groups", len(missing),
+			// Logging all rule groups may excessive, but logging at least 1 may give some hints.
+			"first_missing_rule_group", fmt.Sprintf("user=%s namespace=%s group=%s", missing[0].User, missing[0].Namespace, missing[0].Name))
```

Comment:
i'm not sure about nesting logfmt kv pairs. It's harder to parse in loki (`first_missing_rule_group="user=123 namespace=x group=abc"`)

can we just have separate fields for each?
```suggestion
			"user", missing[0].User,
			"first_missing_rule_group_namespace", missing[0].Namespace,
			"first_missing_rule_group_name", missing[0].Name)
```

Reply:
Very good point. Done in https://github.com/grafana/mimir/pull/9936/commits/776ffccf38dcc9c4a07a8e7bef38efc053da50df